### PR TITLE
network: add support to RoutingPolicyRule lookup table name

### PR DESCRIPTION
**Pull request by ssahani**
Closes https://github.com/systemd/systemd/issues/8418.
Closes https://github.com/systemd/systemd/issues/18121.

### CODE REVIEW

**Comment by bluca** [comment]
Computer says no:

```
../src/network/networkd-route.c:2380:44: error: unused variable 'word' [-Werror,-Wunused-variable]
        _cleanup_free_ char *name = NULL, *word = NULL;
                                           ^
```

**Comment by ssahani** [comment]
Updated thanks for the review. ptal.

**Comment by bluca** [comment]
Computer still says no:

```
../man/networkd.conf.xml
../man/networkd.conf.xml:78: parser error : Opening and ending tag mismatch: literal line 76 and para
ified, then all options specified earlier are cleared. Defaults to unset.</para>
                                                                               ^
../man/networkd.conf.xml:78: parser error : Opening and ending tag mismatch: para line 76 and listitem
n all options specified earlier are cleared. Defaults to unset.</para></listitem
                                                                               ^
../man/networkd.conf.xml:79: parser error : Opening and ending tag mismatch: listitem line 76 and varlistentry
      </varlistentry>
                     ^
../man/networkd.conf.xml:81: parser error : Opening and ending tag mismatch: varlistentry line 76 and variablelist
    </variablelist>
                   ^
../man/networkd.conf.xml:82: parser error : Opening and ending tag mismatch: variablelist line 76 and refsect1
  </refsect1>
             ^
../man/networkd.conf.xml:186: parser error : Opening and ending tag mismatch: refsect1 line 76 and refentry
</refentry>
           ^
../man/networkd.conf.xml:187: parser error : EndTag: '</' not found

^
unable to parse ../man/networkd.conf.xml
```

File: src/network/networkd-route.c [hunk 1]
```diff
@@ -2357,6 +2365,75 @@ int config_parse_multipath_route(
...
+                return 0;
+        }
+
+        m = new(RouteTable, 1);
+        if (!m)
+                return log_oom();
+
+        *m = (RouteTable) {
+                .table = table,
+                .name = TAKE_PTR(name),
```
Comment (bluca) [inline src/network/networkd-route.c]: If ```_ordered_hashmap_ensure_allocated``` (called by ```ordered_hashmap_ensure_put```) fails, what happens to ```name``` ?

Comment (bluca) [reply src/network/networkd-route.c]: never mind, it's freed by the cleanup added above

Comment (ssahani) [reply src/network/networkd-route.c]: ouch thanks fixed. I just forgot in between :)

**Comment by ssahani** [comment]
build passed thanks ~

**Review by bluca — APPROVED** [review]
lgtm, setting green label but waiting to give @yuwata a chance to have a look too

**Comment by flo-wer** [comment]
This clearly doesn't solve the mentioned issues. One problem was the missing parsing of `"default", "main", and "local"` that was promised by the man page. The other was a request for the translation of arbitrary values, followed by the suggestion to match iproute2's behavior, i.e. using values from `/etc/iproute2/rt_tables{,.d/*.conf}`.

And following this suggestion is indeed reasonable, as exactly this path has its purpose for at least [17 years](https://github.com/shemminger/iproute2/blob/aba5acdfdb347d2c21fc67d613d83d4430ca3937/lib/rt_names.c#L274). So instead of duplicating any effort for problems that have been solved decades ago, please just stick to the [iproute2 code](https://github.com/shemminger/iproute2/blob/6fd53b2a1c9a94066f81806700546f39d368b657/lib/rt_names.c#L377) and reuse their logic.

To do some really beneficial improvements, use the same logic for the other iproute2 constants too: `/etc/iproute2/rt_scopes` for `[Address].Scope=` & `[Route].Scope=`, `/etc/iproute2/rt_protos` for `[Route].Protocol=`, `/etc/iproute2/rt_dsfield` for `[Route].IPServiceType=` & `[RoutingPolicyRule].TypeOfService=`.

**Comment by ssahani** [comment]
> This clearly doesn't solve the mentioned issues. One problem was the missing parsing of `"default", "main", and "local"` that was promised by the man page. The other was a request for the translation of arbitrary values, followed by the suggestion to match iproute2's behavior, i.e. using values from `/etc/iproute2/rt_tables{,.d/*.conf}`.

The`"default", "main", and "local"`  is matter of adding defaults.  What man page you talking about. The user can configure what they want.  

> 
> And following this suggestion is indeed reasonable, as exactly this path has its purpose for at least [17 years](https://github.com/shemminger/iproute2/blob/aba5acdfdb347d2c21fc67d613d83d4430ca3937/lib/rt_names.c#L274). So instead of duplicating any effort for problems that have been solved decades ago, please just stick to the [iproute2 code](https://github.com/shemminger/iproute2/blob/6fd53b2a1c9a94066f81806700546f39d368b657/lib/rt_names.c#L377) and reuse their logic.
> 
> To do some really beneficial improvements, use the same logic for the other iproute2 constants too: `/etc/iproute2/rt_scopes` for `[Address].Scope=` & `[Route].Scope=`, `/etc/iproute2/rt_protos` for `[Route].Protocol=`, `/etc/iproute2/rt_dsfield` for `[Route].IPServiceType=` & `[RoutingPolicyRule].TypeOfService=`.

> This clearly doesn't solve the mentioned issues. One problem was the missing parsing of `"default", "main", and "local"` that was promised by the man page. The other was a request for the translation of arbitrary values, followed by the suggestion to match iproute2's behavior, i.e. using values from `/etc/iproute2/rt_tables{,.d/*.conf}`.
> 
> And following this suggestion is indeed reasonable, as exactly this path has its purpose for at least [17 years](https://github.com/shemminger/iproute2/blob/aba5acdfdb347d2c21fc67d613d83d4430ca3937/lib/rt_names.c#L274). So instead of duplicating any effort for problems that have been solved decades ago, please just stick to the [iproute2 code](https://github.com/shemminger/iproute2/blob/6fd53b2a1c9a94066f81806700546f39d368b657/lib/rt_names.c#L377) and reuse their logic.

It does not make any sense to duplicate their code.  
> 
> To do some really beneficial improvements, use the same logic for the other iproute2 constants too: `/etc/iproute2/rt_scopes` for `[Address].Scope=` & `[Route].Scope=`, `/etc/iproute2/rt_protos` for `[Route].Protocol=`, `/etc/iproute2/rt_dsfield` for `[Route].IPServiceType=` & `[RoutingPolicyRule].TypeOfService=`.

Can you raise RFE what precisely you want for separate feature.

**Comment by bluca** [comment]
Removed green label as there were some changes, will review again but later today

**Comment by ssahani** [comment]
> Removed green label as there were some changes, will review again but later today

Yeah I added a supporting commit for routing and by default allowed main/default etc. thanks

File: src/network/networkd-route.c [hunk 2]
```diff
@@ -2355,6 +2372,75 @@ int config_parse_multipath_route(
...
+        };
+
+        r = ordered_hashmap_ensure_put(s, &string_hash_ops, m->name, m);
```
Comment (yuwata) [inline src/network/networkd-route.c]: Why do you use `ordered_hashmap`? I think simple `hashmap` is enough.
Also, It is not necessary to introduce `struct RoutaTable`. But just store route table number with its name as a key:
```
r = hashmap_ensure_put(s, &string_hash_ops, name, UINT32_TO_PTR(table));
```

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by ssahani** [comment]
Updated. thanks. ptal

File: src/network/networkd-manager.c [hunk 3]
```diff
@@ -875,6 +875,8 @@ void manager_free(Manager *m) {
 
         ordered_set_free_free(m->address_pools);
 
+        m->route_tables = hashmap_free_free(m->route_tables);
```
Comment (yuwata) [inline src/network/networkd-manager.c]: Please use `hashmap_free_free_key()`.

File: src/network/networkd-route.c [hunk 4]
```diff
@@ -2356,6 +2362,67 @@ int config_parse_multipath_route(
...
+        const char *p;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (isempty(rvalue)) {
+                *s = hashmap_free_free(*s);
```
Comment (yuwata) [inline src/network/networkd-route.c]: `hashmap_free_free_key()`

File: src/network/networkd-route.c [hunk 5]
```diff
@@ -2356,6 +2362,67 @@ int config_parse_multipath_route(
...
+
+        r = safe_atou32(p, &table);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse RouteTable=, ignoring assignment: %s", p);
+                return 0;
+        }
+
+        if (table == 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
```
Comment (yuwata) [inline src/network/networkd-route.c]: `r` -> `0`.

File: src/network/networkd-route.c [hunk 6]
```diff
@@ -2356,6 +2362,67 @@ int config_parse_multipath_route(
...
+                           "Invalid RouteTable=, ignoring assignment: %s", p);
+                return 0;
+        }
+
+        r = hashmap_ensure_put(s, &string_hash_ops, name, UINT32_TO_PTR(table));
+        if (r == -ENOMEM)
+                return log_oom();
+
```
Comment (yuwata) [inline src/network/networkd-route.c]: Not a matter, but I'd like to drop this blank line.

File: src/network/networkd-route.c [hunk 7]
```diff
@@ -2356,6 +2362,67 @@ int config_parse_multipath_route(
...
+        if (r < 0)
+                return log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to store RouteTable=, ignoring: %s", rvalue);
```
Comment (yuwata) [inline src/network/networkd-route.c]:
```
if (r < 0) {
    log_syntax(...);
    return 0;
}
if (r == 0) {
    log_syntax(..., 0, "Specified route table name and value pair %s is duplicated, ignoring assignment: %s", rvalue);
    return 0;
}

TAKE_PTR(name);
```

File: src/network/networkd-route.c [hunk 8]
```diff
@@ -1903,11 +1903,17 @@ int config_parse_route_table(
...
+                        uint32_t t = PTR_TO_UINT32(hashmap_get(network->manager->route_tables, rvalue));
+                        if (t)
+                                n->table = t;
+                } else {
+                        r = safe_atou32(rvalue, &n->table);
+                        if (r < 0) {
+                                log_syntax(unit, LOG_WARNING, filename, line, r,
+                                           "Could not parse route table number \"%s\", ignoring assignment: %m", rvalue);
+                                return 0;
+                        }
```
Comment (yuwata) [inline src/network/networkd-route.c]: This block is duplicated in networkd-routing-policy-rule.c. How about to introduce `route_table_from_string_full()` or something?

File: man/networkd.conf.xml [hunk 9]
```diff
@@ -70,6 +70,14 @@
         is false. Defaults to yes.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>RouteTable=</varname></term>
+        <listitem><para>Specifies the route table name. Takes a route name and table number separated with a colon.
+        (<literal><replaceable>name</replaceable>:<replaceable>integer</replaceable></literal>. The route table number
+        must be an integer in the range 1..4294967295. This setting can be specified multiple times. If an empty string
+        is specified, then all options specified earlier are cleared. Defaults to unset.</para></listitem>
```
Comment (yuwata) [inline man/networkd.conf.xml:78]: I'd like to this accepts a white-space separated list of name:integer pairs, eg `RouteTable=hoge:123 foo:456`.

**Review by yuwata — CHANGES_REQUESTED** [review]
Please do not mix the change in the first and second commits, but separate changes clearly.

File: man/networkd.conf.xml (hunk 9 above)
Comment (ssahani) [reply man/networkd.conf.xml:78]: This is OK for few tables but what about `4294967295`  number of tables and imposible to write in one line . Hence This is this way

File: src/network/networkd-route.c (hunk 8 above)
Comment (ssahani) [reply src/network/networkd-route.c]: yes I was think about it

File: man/networkd.conf.xml (hunk 9 above)
Comment (yuwata) [reply man/networkd.conf.xml:78]: Of course, the setting should _also_ accept multiple assignments, e.g.
```
RouteTable=hoge:123 foo:456
RouteTable=aaa:1111 bbb:2222 ccc:3333
```

Comment (ssahani) [reply man/networkd.conf.xml:78]: It's too complex . Let's just stick to same as dhcp options much easier to manage for the user also.

Comment (yuwata) [reply man/networkd.conf.xml:78]: I do not think so, e.g. `SystemCallFilter=`. But ok, let's extend that in another PR.

**Comment by ssahani** [comment]
updated thanks ptal

File: src/network/networkd-route.c [hunk 10]
```diff
@@ -1868,6 +1868,31 @@ int config_parse_route_scope(
...
+int route_table_from_string_full(Network *network, const char *s, uint32_t *ret) {
+        int r;
+
+        assert(s);
+        assert(network);
+        assert(ret);
+
+        r = route_table_from_string(s);
+        if (r >= 0)
+                *ret = (uint32_t) r;
```
Comment (yuwata) [inline src/network/networkd-route.c]: To reduce the indent below, how about `return` in this block?
```
if (r >= 0) {
    *ret = (uint32_t) r;
    return 0;
}
```

File: src/network/networkd-route.c [hunk 11]
```diff
@@ -1868,6 +1868,31 @@ int config_parse_route_scope(
...
+        else {
+                if (!hashmap_isempty(network->manager->route_tables)) {
```
Comment (yuwata) [inline src/network/networkd-route.c]: If the hashmap is not empty, then the setting cannot accept integer... So, please drop `hashmap_isempty()` and do like the following:
```
uint32_t t = PTR_TO_UINT32(hashmap_get(...));
if (t != 0) {
    *ret = t;
    return 0;
}

return safe_atou32(s, ret);
```

File: src/network/networkd-route.c [hunk 12]
```diff
@@ -2356,6 +2376,67 @@ int config_parse_multipath_route(
...
+                           "Invalid RouteTable=, ignoring assignment: %s", p);
+                return 0;
+        }
+
+        r = hashmap_ensure_put(s, &string_hash_ops, name, UINT32_TO_PTR(table));
+        if (r == -ENOMEM)
+                return log_oom();
+        if (r < 0)
+                return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                  "Specified RouteTable= name and value pair is duplicated, ignoring assignment: %s", rvalue);
```
Comment (yuwata) [inline src/network/networkd-route.c]: `hashmap_ensure_put()` returns `-ENOMEM`, `-EEXIST`, `0`, or `1`. So, please adjust the log message here:
`"Specified RouteTable= name and value pair conflicts with others, ignoring assignment: %s"`

File: src/network/networkd-route.c [hunk 13]
```diff
@@ -2356,6 +2376,67 @@ int config_parse_multipath_route(
...
+
```
Comment (yuwata) [inline src/network/networkd-route.c:2443]: Please do not free `name`:
```
if (r > 0)
    TAKE_PTR(name);
```

File: src/network/networkd-route.c [hunk 14]
```diff
@@ -1868,6 +1868,31 @@ int config_parse_route_scope(
         return 0;
 }
 
+int route_table_from_string_full(Network *network, const char *s, uint32_t *ret) {
```
Comment (yuwata) [inline src/network/networkd-route.c]: I'd like this to take `Manager *` instead of `Network *`.

File: man/systemd.network.xml [hunk 15]
```diff
@@ -1202,7 +1202,9 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
           <listitem>
             <para>Specifies the routing table identifier to lookup if the rule selector matches. Takes
             one of <literal>default</literal>, <literal>main</literal>, and <literal>local</literal>,
-            or a number between 1 and 4294967295. Defaults to <literal>main</literal>.</para>
+            and a number between 1 and 4294967295 or specified in <varname>RouteTable=</varname>.
```
Comment (yuwata) [inline man/systemd.network.xml]:
```
Takes one of predefined names <literal>default</literal>, <literal>main</literal>, and <literal>local</literal>,
and names defined in <varname>RouteTable=</varname> in
<citerefentry><refentrytitle>networkd.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>,
or a number between 1 and 4294967295.
```

File: man/systemd.network.xml [hunk 16]
```diff
@@ -1410,7 +1412,8 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
           <term><varname>Table=</varname></term>
           <listitem>
             <para>The table identifier for the route. Takes <literal>default</literal>,
-            <literal>main</literal>, <literal>local</literal> or a number between 1 and 4294967295.
+            <literal>main</literal>, <literal>local</literal> and a number between 1 and 4294967295 or specified in
+            <varname>RouteTable=</varname>. See <citerefentry><refentrytitle>networkd.conf</refentrytitle><manvolnum>5</manvolnum></citerefentry>
```
Comment (yuwata) [inline man/systemd.network.xml]: Same here.

File: src/network/networkd-route.c [hunk 17]
```diff
@@ -2356,6 +2376,67 @@ int config_parse_multipath_route(
...
+                           "Failed to parse RouteTable=, ignoring assignment: %s", p);
+                return 0;
+        }
+
+        if (table == 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                           "Invalid RouteTable=, ignoring assignment: %s", p);
+                return 0;
+        }
+
```
Comment (yuwata) [inline src/network/networkd-route.c:2432]: Maybe, drop the case that `name` is `"default"`, `"main"`, or `"local"`.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/network/networkd-route.c (hunk 17 above)
Comment (ssahani) [reply src/network/networkd-route.c:2432]: It's ok if user wants them there.

**Comment by ssahani** [comment]
updated. thanks for the review

File: src/network/networkd-route.c [hunk 18]
```diff
@@ -2356,6 +2373,69 @@ int config_parse_multipath_route(
...
+        r = hashmap_ensure_put(s, &string_hash_ops, name, UINT32_TO_PTR(table));
+        if (r == -ENOMEM)
+                return log_oom();
+        if (r == -EEXIST)
+                return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                  "Specified RouteTable= name and value pair conflicts with others, ignoring assignment: %s", rvalue);
```
Comment (yuwata) [inline src/network/networkd-route.c]: Nah, please do not return negative errno, but _ignore_ the error:
```
if (r < 0) {
    log_syntax(...);
    return 0;
}

File: src/network/networkd-route.c (hunk 17 above)
Comment (yuwata) [reply src/network/networkd-route.c:2432]: Hmm, but in `route_table_from_string_full()` we first check the string is one of `default`, `main` or `local`. So, setting `RouteTable=default:12345` is completely ignored without any warning...

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/network/networkd-route.c (hunk 17 above)
Comment (ssahani) [reply src/network/networkd-route.c:2432]: We will just give a warning log

Comment (ssahani) [reply src/network/networkd-route.c:2432]: like 
```
etc/systemd/networkd.conf:3: Route table name local alreay preconfigured. Ignoring assignment: local:255
/etc/systemd/networkd.conf:4: Route table name main alreay preconfigured. Ignoring assignment: main:254
/etc/systemd/networkd.conf:5: Route table name default alreay preconfigured. Ignoring assignment: default:253
```

**Comment by yuwata** [comment]
Thanks. LGTM.

**Review by yuwata — APPROVED** [review]

File: src/network/networkd-route.c [hunk 19]
```diff
@@ -2356,6 +2373,77 @@ int config_parse_multipath_route(
...
+        if (r <= 0 || isempty(p)) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Invalid RouteTable=, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        if (STR_IN_SET(name, "default", "main","local")) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Route table name %s already preconfigured. Ignoring assignment: %s", name, rvalue);
+                return 0;
```
Comment (ssahani) [inline src/network/networkd-route.c:2417]: fixed the spelling mistake here

**Comment by bluca** [comment]
bionic-ppc64 fails in TEST-53-ISSUE-16347 which is to do with timer units, looks unrelated

**Comment by glance-** [comment]
I'd love to see this work (or something like #18275) finished and merged. How can I help to make that happen?

**Pull request merged**